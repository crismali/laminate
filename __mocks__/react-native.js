/* eslint-env jest */

class Value {
  constructor (value) {
    this.value = value
  }

  interpolate (args) {
    return { interpolated: this.value, args: args }
  }
}

const AnimatedView = ({ children }) => children
AnimatedView.displayName = 'Animated.View'

const animatedTimingStart = jest.fn()
const Animated = {
  Value,
  View: AnimatedView,
  timing: jest.fn(() => ({ start: animatedTimingStart })),
  animatedTimingStart
}

const Dimensions = { get: jest.fn(() => ({ width: 375, height: 667 })) }

const PanResponder = { create: jest.fn((args) => ({ panResponder: args })) }

const View = ({ children }) => children

const StyleSheet = { create: jest.fn(styles => styles) }

export {
  Animated,
  Dimensions,
  PanResponder,
  StyleSheet,
  View
}
