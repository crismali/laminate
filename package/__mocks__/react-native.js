/* eslint-env jest */
import React from 'react'

class Value {
  constructor (value) {
    this.value = value
  }

  interpolate (args) {
    return { interpolated: this.value, args: args }
  }

  setValue (value) {
    this.value = value
  }
}

const AnimatedView = ({ children }) => children
AnimatedView.displayName = 'Animated.View'

const animatedTimingStart = jest.fn((func) => func && func())
const Animated = {
  Value,
  View: AnimatedView,
  timing: jest.fn(() => ({ start: animatedTimingStart })),
  animatedTimingStart
}

const Dimensions = { get: jest.fn(() => ({ width: 375, height: 667 })) }

const Easing = { inOut: jest.fn() }

const PanResponder = { create: jest.fn((args) => ({ panHandlers: args })) }

const View = ({ children }) => <div>{children}</div> // eslint-disable-line

const StyleSheet = { create: jest.fn(styles => styles) }

export {
  Animated,
  Dimensions,
  Easing,
  PanResponder,
  StyleSheet,
  View
}
