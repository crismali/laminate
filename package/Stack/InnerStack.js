import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { View, StyleSheet, Text } from 'react-native'
import { Animation } from '../Animation'
import { SceneWrapper } from './SceneWrapper'

const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: '100%'
  }
})

const allButLast = (collection) => collection.slice(0, collection.length - 1)

export class InnerStack extends Component {
  static propTypes = {
    ...Animation.childContextTypes,
    children: PropTypes.any.isRequired,
    innerRouter: PropTypes.object.isRequired,
    outerRouter: PropTypes.object.isRequired
  }

  static childContextTypes = {
    outerRouter: PropTypes.object.isRequired
  }

  state = {
    scenes: [this.props.children]
  }

  getChildContext () {
    return {
      outerRouter: this.props.outerRouter
    }
  }

  componentDidMount () {
    this.props.animation.play()
  }

  componentWillReceiveProps (nextProps) {
    const { innerRouter: newInnerRouter } = nextProps
    const scenesLength = this.scenes().length
    const { index: historyIndex } = newInnerRouter.history

    switch (newInnerRouter.history.action) {
      case 'REPLACE':
        this.handleReplace(nextProps); break
      case 'PUSH':
        if (scenesLength !== historyIndex + 1) {
          this.handlePush(nextProps)
        }
        break
      case 'POP':
        this.handlePop(nextProps)
        break
    }
  }

  scenes = () => this.state.scenes

  handlePop = (nextProps) => {
    const scenes = this.scenes().slice(0, nextProps.innerRouter.history.index + 1)
    this.props.animation.rewind({
      toValue: scenes.length,
      callback: () => this.setState({ scenes })
    })
  }

  handlePush = (nextProps) => {
    const scenes = [...this.scenes(), nextProps.children]
    this.setState({ scenes })
    setTimeout(() => nextProps.animation.play({ toValue: scenes.length }), 10)
  }

  handleReplace = (nextProps) => {
    const scenes = [...allButLast(this.scenes()), nextProps.children]
    this.setState({ scenes })
  }

  render () {
    return <View style={styles.container}>
      {this.scenes().map((scene, index) => (
        <SceneWrapper
          key={index}
          index={index}
          animationValue={this.props.animation.value}
        >
          {scene}
        </SceneWrapper>
      ))}
    </View>
  }
}
