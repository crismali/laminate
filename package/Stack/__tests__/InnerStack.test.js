/* eslint-env jest */
import React from 'react'
import { Animated, Text } from 'react-native'
import { shallow } from 'enzyme'
import { uniqueId } from 'lodash'
import { InnerStack } from '../InnerStack'

const Child = () => <Text>a child</Text>

const animationMock = () => {
  return {
    play: jest.fn(),
    rewind: jest.fn(({ callback }) => callback && callback()),
    value: new Animated.Value(17)
  }
}

const routerMock = (history) => {
  return {
    history: history || {},
    location: { key: uniqueId(), pathname: '/foo' },
    match: {}
  }
}

describe('state', () => {
  it('has scenes', () => {
    const children = <Child />
    const subject = shallow(<InnerStack animation={animationMock()} router={routerMock()}>
      {children}
    </InnerStack>)

    expect(subject.instance().state.scenes).toEqual([children])
  })
})

describe('#componentDidMount', () => {
  it('plays the animation', () => {
    const animation = animationMock()
    const subject = shallow(<InnerStack animation={animation} router={routerMock()}>
      <Child />
    </InnerStack>)
    subject.instance().componentDidMount()
    expect(animation.play).toHaveBeenCalled()
  })
})

describe('#componentWillReceiveProps', () => {
  let subject
  let children

  beforeEach(() => {
    children = <Child>
      <Child path='/foo' />
      <Child />
      <Child />
    </Child>
    const animation = animationMock()
    subject = shallow(<InnerStack animation={animation} router={routerMock()}>
      {children}
    </InnerStack>)
    subject.instance().handleReplace = jest.fn()
    subject.instance().handlePush = jest.fn()
    subject.instance().handlePop = jest.fn()
  })

  it('handles replacing', () => {
    const nextProps = {
      router: routerMock({ action: 'REPLACE', index: 2 }),
      children
    }
    subject.instance().componentWillReceiveProps(nextProps)
    expect(subject.instance().handleReplace).toHaveBeenCalledWith(nextProps)
    expect(subject.instance().handlePush).not.toHaveBeenCalled()
    expect(subject.instance().handlePop).not.toHaveBeenCalled()
  })

  it('handles pushing', () => {
    const nextProps = {
      router: routerMock({ action: 'PUSH' }),
      children
    }
    subject.instance().componentWillReceiveProps(nextProps)
    expect(subject.instance().handlePush).toHaveBeenCalledWith(nextProps)
    expect(subject.instance().handleReplace).not.toHaveBeenCalled()
    expect(subject.instance().handlePop).not.toHaveBeenCalled()
  })

  it('handles popping', () => {
    const nextProps = {
      router: routerMock({ action: 'POP' }),
      children
    }
    subject.instance().componentWillReceiveProps(nextProps)
    expect(subject.instance().handlePop).toHaveBeenCalledWith(nextProps)
    expect(subject.instance().handlePush).not.toHaveBeenCalled()
    expect(subject.instance().handleReplace).not.toHaveBeenCalled()
  })
})

describe('#handlePop', () => {
  it('rewinds the animation and removes scenes from the stack based on the history index', () => {
    const animation = animationMock()
    const subject = shallow(<InnerStack animation={animation} router={routerMock()}>
      <Child />
    </InnerStack>)
    const nextProps = {
      router: routerMock({ action: 'POP' })
    }
    subject.setState({ scenes: [<Child key={1} />, <Child key={2} />, <Child key={3} />] })
    subject.instance().handlePop(nextProps)
    expect(animation.rewind).toHaveBeenCalledWith({ toValue: 2, callback: expect.any(Function) })
    expect(subject.instance().state.scenes).toEqual([<Child key={1} />, <Child key={2} />])
  })
})

describe('#handlePush', () => {
  it('adds a scene to the stack and plays the animation', () => {
    const animation = animationMock()
    const subject = shallow(<InnerStack animation={animation} router={routerMock()}>
      <Child key='original' />
    </InnerStack>)
    const nextProps = {
      animation,
      children: <Child key={17} />,
      router: { history: { action: 'PUSH', index: 1 } }
    }
    subject.instance().handlePush(nextProps)
    expect(subject.instance().state.scenes).toEqual([<Child key='original' />, <Child key={17} />])
    expect(animation.play).toHaveBeenCalledWith({ toValue: 2 })
  })
})

describe('#handleReplace', () => {
  it('replaces the last scene in the stack and plays the animation', () => {
    const animation = animationMock()
    const subject = shallow(<InnerStack animation={animation} router={routerMock()}>
      <Child key='original' />
    </InnerStack>)
    const nextProps = {
      animation,
      children: <Child key={17} />,
      router: { history: { action: 'REPLACE', index: 0 } }
    }
    subject.instance().handleReplace(nextProps)
    subject.update()
    expect(subject.instance().state.scenes).toEqual([<Child key={17} />])
    expect(animation.play).toHaveBeenCalledWith({ toValue: 1 })
  })
})

describe('#render', () => {
  it('renders the scenes properly', () => {
    const animation = animationMock()
    const subject = shallow(<InnerStack animation={animation} router={routerMock()}>
      <Child key='original' />
    </InnerStack>)
    subject.setState({
      scenes: [<Child key={1} foo={1} />, <Child key={2} foo={2} />, <Child key={3} foo={3} />, <Child key={4} foo={4} />]
    })
    expect(subject).toMatchSnapshot()
  })
})
