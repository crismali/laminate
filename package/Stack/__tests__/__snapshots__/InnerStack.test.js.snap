// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render renders the scenes properly 1`] = `
Array [
  <SceneWrapper
    animationValue={
      Value {
        "value": 17,
      }
    }
    index={0}
    key="0"
  />,
  <SceneWrapper
    animationValue={
      Value {
        "value": 17,
      }
    }
    index={1}
    key="1"
  />,
  <SceneWrapper
    animationValue={
      Value {
        "value": 17,
      }
    }
    index={2}
    key="2"
  />,
  <SceneWrapper
    animationValue={
      Value {
        "value": 17,
      }
    }
    index={3}
    key="3"
  />,
]
`;
